skrifa/src/outline/glyf/hint/engine/outline.rs

```diff
@@ -360,6 +360,412 @@ impl<'a> Engine<'a> {
+        let gs = &mut self.graphics_state;
+        let distance = F26Dot6::from_bits(
+            gs.project(gs.zp0().point(p2)?, gs.zp1().point(p1)?)
+                .to_bits()
+                / 2,
+        );
+        gs.move_point(gs.zp1, p1, distance)?;
+        gs.move_point(gs.zp0, p2, -distance)?;
+        Ok(())
+    }
+
+    /// Interpolate point by last relative stretch.
+    ///
+    /// IP[] (0x39)
+    ///
+    /// Pops: p: point number
+    ///
+    /// Uses the loop counter.
+    ///
+    /// Moves point p so that its relationship to rp1 and rp2 is the same as it
+    /// was in the original uninstructed outline. Measurements are made along
+    /// the projection_vector, and movement to satisfy the interpolation
+    /// relationship is constrained to be along the freedom_vector. This
+    /// instruction is not valid if rp1 and rp2 have the same position on the
+    /// projection_vector.
+    ///
+    /// See <https://learn.microsoft.com/en-us/typography/opentype/spec/tt_instructions#interpolate-point-by-the-last-relative-stretch>
+    /// and <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/truetype/ttinterp.c#L6065>
+    pub(super) fn op_ip(&mut self) -> OpResult {
+        let gs = &mut self.graphics_state;
+        let in_twilight = gs.zp0.is_twilight() || gs.zp1.is_twilight() || gs.zp2.is_twilight();
+        let orus_base = if in_twilight {
+            gs.zp0().original(gs.rp1)?
+        } else {
+            gs.zp0().unscaled(gs.rp1)?.map(F26Dot6::from_bits)
+        };
+        let cur_base = gs.zp0().point(gs.rp1)?;
+        let old_range = if in_twilight {
+            gs.dual_project(gs.zp1().original(gs.rp2)?, orus_base)
+        } else {
+            gs.dual_project(
+                gs.zp1().unscaled(gs.rp2)?.map(F26Dot6::from_bits),
+                orus_base,
+            )
+        };
+        let cur_range = gs.project(gs.zp1().point(gs.rp2)?, cur_base);
+        let count = gs.loop_counter;
+        gs.loop_counter = 1;
+        for _ in 0..count {
+            let point = self.value_stack.pop_usize()?;
+            let original_distance = if in_twilight {
+                gs.dual_project(gs.zp2().original(point)?, orus_base)
+            } else {
+                gs.dual_project(gs.zp2().unscaled(point)?.map(F26Dot6::from_bits), orus_base)
+            };
+            let cur_distance = gs.project(gs.zp2().point(point)?, cur_base);
+            let mut new_distance = F26Dot6::ZERO;
+            if original_distance != F26Dot6::ZERO {
+                if old_range != F26Dot6::ZERO {
+                    new_distance = F26Dot6::from_bits(math::mul_div(
+                        original_distance.to_bits(),
+                        cur_range.to_bits(),
+                        old_range.to_bits(),
+                    ));
+                } else {
+                    new_distance = original_distance;
+                }
```

Comment:
there are a few of these where I would find `let val = if cond` to be slightly preferable to mutation? but 🤷

Reply:
Yeah :/ I tried not to stray too far from FreeType code but I can fix some of these up.